jmdaemon/jmdaemon/orderbookwatch.py

```diff
@@ -3,10 +3,13 @@
 import sqlite3
 import sys
 import threading
+import base64
+import struct
 from decimal import InvalidOperation, Decimal
 from numbers import Integral
 
 from jmdaemon.protocol import JM_VERSION
+import jmbitcoin as btc
```

Comment:
In fact this is a bigger issue with fidelity bonds as implemented right now. Dependency on the `jmbitcoin` package in `jmdaemon` was also added by the other code. imo the twisted protocol should be extended so the daemon loops back to the client to receive a signed fidelity bonds proof message on orderbook requests. Right now jmdaemon is handed the private key data and creates the signed proof itself.

Reply:
> Dependency on the `jmbitcoin` package in `jmdaemon` was also added by the other code.

Ah yes, I hadn't noticed that yet. Indeed, the certificate key signing on `taker_nick|maker_nick` is done in the daemon. (Although I still don't know why the FB proof contains the utxo privkey *as well as* the cert privkey? Need to read again, but whether there's a good reason or not, that'll need to be removed, too).

>  imo the twisted protocol should be extended so the daemon loops back to the client to receive a signed fidelity bonds proof message on orderbook requests.

Yes this is one of the 'ways to address' this I was mentioning in my last paragraph, and indeed, it is, as described, the current way we handle the same problem for message signing on the message channel. It's a shame to have that extra 2 rounds of communication, but it's somewhat needed.

Originally I was thinking that there might be easier ways to handle the problem here: you could:

* defer the signature checks until the data gets back to `jmclient` and address the potential for the orderbook getting spammed up with invalid data by having an explicit `DELETE` message going back from client to daemon
* or: since we currently have no defence against that spam, maybe don't bother with the DELETE part.

However as you point out, we also must address this issue in the sign/create fidelity bond workflow, and there, since the message to be signed is dynamic, we have to consider another choice:
* Either copy the current paradigm of `jmbase.commands.JMRequestMsgSig` and `jmbase.commands.JMRequestMsgSigVerify` messages sendable from daemon to client (hmm actually maybe we can literally use the same AMP commands? But be careful! There is some trickiness in that code)
* Or: more "out there" but arguably more elegant: `jmdaemon` currently uses libsodium for authenticated encryption; it also has signature functionality; maybe we can just make the certificate key/signature be using that crypto library? Tbh I don't like this but thought I'd mention it.